### ISSUE TITLE
Report disjoint entities with data model

### ISSUE THREAD

**Issue by bruth** [PEDSnet/tools#37]
By default, only the fields flagged `required` will be reported with the option of setting the `verbose` flag to see all missing entities. Further entities that no longer exist in the data model will be reported.

**Comment by murphyke** [comment]
If we want to test the tool on the submissions we have so far, I've thrown them together in the [prov-data branch of Data_Coordinating_Center](https://github.com/PEDSnet/Data_Coordinating_Center/tree/prov-data/provenance/data)

**Comment by bruth** [comment]
That is great. Thanks.

**Comment by bruth** [comment]
@murphyke It appears the operational tables such as [concept](https://github.com/chop-dbhi/data-models/blob/master/pedsnet/v2/definitions/concept.csv) are marked as required so I added a fixed list of PEDSnet tables in the validator to produce the results below.

**Boston**

The files validated without issue against 1.0.0. It does not seem they have been updated to the 2.0.0 data model.

```
pedsnet-etlprov -model pedsnet -version 2.0.0 boston
Validating against model 'pedsnet/2.0.0'
A few problems have been detected with the entities.csv file:
* Unknown field 'care_site.organization_id'
* Unknown field 'condition_occurrence.associated_provider_id'
* Unknown field 'death.cause_of_death_concept_id'
* Unknown field 'death.cause_of_death_source_value'
* Unknown field 'observation.associated_provider_id'
* Unknown field 'observation.range_high'
* Unknown field 'observation.range_low'
* Unknown field 'observation.relevant_condition_concept_id'
* Unknown field 'person.pn_time_of_birth'
* Unknown field 'procedure_occurrence.associated_provider_id'
* Unknown field 'procedure_occurrence.relevant_condition_concept_id'
* Unknown field 'visit_occurrence.place_of_service_concept_id'
* Unknown field 'visit_occurrence.place_of_service_source_value'
* Missing field 'condition_occurrence.condition_start_time'
* Missing field 'death.death_time'
* Missing field 'drug_exposure.drug_concept_id'
* Missing field 'drug_exposure.drug_exposure_end_time'
* Missing field 'drug_exposure.drug_exposure_id'
* Missing field 'drug_exposure.drug_exposure_start_date'
* Missing field 'drug_exposure.drug_exposure_start_time'
* Missing field 'drug_exposure.drug_type_concept_id'
* Missing field 'drug_exposure.person_id'
* Missing field 'fact_relationship.domain_concept_id_1'
* Missing field 'fact_relationship.domain_concept_id_2'
* Missing field 'fact_relationship.fact_id_1'
* Missing field 'fact_relationship.fact_id_2'
* Missing field 'fact_relationship.relationship_concept_id'
* Missing field 'measurement.measurement_concept_id'
* Missing field 'measurement.measurement_date'
* Missing field 'measurement.measurement_id'
* Missing field 'measurement.measurement_source_value'
* Missing field 'measurement.measurement_type_concept_id'
* Missing field 'measurement.person_id'
* Missing field 'measurement.unit_source_value'
* Missing field 'measurement.value_source_value'
* Missing field 'observation_period.observation_period_start_time'
* Missing field 'observation_period.period_type_concept_id'
* Missing field 'procedure_occurrence.procedure_time'
* Missing field 'visit_occurrence.visit_concept_id'
* Missing field 'visit_occurrence.visit_type_concept_id'
* Missing field 'visit_payer.plan_class'
* Missing field 'visit_payer.plan_name'
* Missing field 'visit_payer.visit_occurrence_id'
* Missing field 'visit_payer.visit_payer_id'
File steps.csv looks good!
File tools.csv looks good!
File sources.csv looks good!
File people.csv looks good!
Found 93 entities
Found 4 steps
Found 2 tools
Found 3 sources
Found 1 persons
```

**CHOP**

```
pedsnet-etlprov -model pedsnet -version 2.0.0 chop
Validating against model 'pedsnet/2.0.0'
A few problems have been detected with the entities.csv file:
* Missing field 'visit_occurrence.visit_concept_id'
File steps.csv looks good!
File tools.csv looks good!
File sources.csv looks good!
File people.csv looks good!
Found 182 entities
Found 133 steps
Found 1 tools
Found 2 sources
Found 1 persons
```

**Nationwide**

```
pedsnet-etlprov -model i2b2_pedsnet -version 2.0.0 nationwide
Validating against model 'i2b2_pedsnet/2.0.0'
A few problems have been detected with the entities.csv file:
* Missing field 'provider_dimension.provider_id'
* Missing field 'provider_dimension.provider_path'
File steps.csv looks good!
File tools.csv looks good!
File sources.csv looks good!
File people.csv looks good!
Found 101 entities
Found 25 steps
Found 15 tools
Found 2 sources
Found 3 persons
```

**Nemours**

Nemours files did not validate against 2.0.0 nor 1.0.0.

```
pedsnet-etlprov -model pedsnet -version 2.0.0 nemours
Validating against model 'pedsnet/2.0.0'
A few problems have been detected with the entities.csv file:
* Missing field 'care_site.care_site_id'
* Missing field 'care_site.care_site_source_value'
* Missing field 'condition_occurrence.condition_concept_id'
* Missing field 'condition_occurrence.condition_occurrence_id'
* Missing field 'condition_occurrence.condition_start_date'
* Missing field 'condition_occurrence.condition_start_time'
* Missing field 'condition_occurrence.condition_type_concept_id'
* Missing field 'condition_occurrence.person_id'
* Missing field 'death.death_date'
* Missing field 'death.death_time'
* Missing field 'death.death_type_concept_id'
* Missing field 'death.person_id'
* Missing field 'drug_exposure.drug_concept_id'
* Missing field 'drug_exposure.drug_exposure_end_time'
* Missing field 'drug_exposure.drug_exposure_id'
* Missing field 'drug_exposure.drug_exposure_start_date'
* Missing field 'drug_exposure.drug_exposure_start_time'
* Missing field 'drug_exposure.drug_type_concept_id'
* Missing field 'drug_exposure.person_id'
* Missing field 'fact_relationship.domain_concept_id_1'
* Missing field 'fact_relationship.domain_concept_id_2'
* Missing field 'fact_relationship.fact_id_1'
* Missing field 'fact_relationship.fact_id_2'
* Missing field 'fact_relationship.relationship_concept_id'
* Missing field 'location.location_id'
* Missing field 'measurement.measurement_concept_id'
* Missing field 'measurement.measurement_date'
* Missing field 'measurement.measurement_id'
* Missing field 'measurement.measurement_source_value'
* Missing field 'measurement.measurement_type_concept_id'
* Missing field 'measurement.person_id'
* Missing field 'measurement.unit_source_value'
* Missing field 'measurement.value_source_value'
* Missing field 'observation.observation_concept_id'
* Missing field 'observation.observation_date'
* Missing field 'observation.observation_id'
* Missing field 'observation.observation_type_concept_id'
* Missing field 'observation.person_id'
* Missing field 'observation_period.observation_period_id'
* Missing field 'observation_period.observation_period_start_date'
* Missing field 'observation_period.observation_period_start_time'
* Missing field 'observation_period.period_type_concept_id'
* Missing field 'observation_period.person_id'
* Missing field 'person.care_site_id'
* Missing field 'person.gender_concept_id'
* Missing field 'person.person_id'
* Missing field 'person.person_source_value'
* Missing field 'person.year_of_birth'
* Missing field 'procedure_occurrence.person_id'
* Missing field 'procedure_occurrence.procedure_concept_id'
* Missing field 'procedure_occurrence.procedure_date'
* Missing field 'procedure_occurrence.procedure_occurrence_id'
* Missing field 'procedure_occurrence.procedure_time'
* Missing field 'procedure_occurrence.procedure_type_concept_id'
* Missing field 'provider.care_site_id'
* Missing field 'provider.provider_id'
* Missing field 'provider.provider_source_value'
* Missing field 'visit_occurrence.person_id'
* Missing field 'visit_occurrence.visit_concept_id'
* Missing field 'visit_occurrence.visit_occurrence_id'
* Missing field 'visit_occurrence.visit_start_date'
* Missing field 'visit_occurrence.visit_type_concept_id'
* Missing field 'visit_payer.plan_class'
* Missing field 'visit_payer.plan_name'
* Missing field 'visit_payer.visit_occurrence_id'
* Missing field 'visit_payer.visit_payer_id'
File steps.csv looks good!
File tools.csv looks good!
File sources.csv looks good!
File people.csv looks good!
Found 95 entities
Found 24 steps
Found 10 tools
Found 2 sources
Found 3 persons
```

```
pedsnet-etlprov -model pedsnet -version 1.0.0 nemours
Validating against model 'pedsnet/1.0.0'
A few problems have been detected with the entities.csv file:
* Missing field 'care_site.care_site_id'
* Missing field 'care_site.care_site_source_value'
* Missing field 'care_site.organization_id'
* Missing field 'condition_occurrence.condition_concept_id'
* Missing field 'condition_occurrence.condition_occurrence_id'
* Missing field 'condition_occurrence.condition_start_date'
* Missing field 'condition_occurrence.condition_type_concept_id'
* Missing field 'condition_occurrence.person_id'
* Missing field 'death.death_date'
* Missing field 'death.death_type_concept_id'
* Missing field 'death.person_id'
* Missing field 'location.location_id'
* Missing field 'observation.observation_concept_id'
* Missing field 'observation.observation_date'
* Missing field 'observation.observation_id'
* Missing field 'observation.observation_type_concept_id'
* Missing field 'observation.person_id'
* Missing field 'observation_period.observation_period_id'
* Missing field 'observation_period.observation_period_start_date'
* Missing field 'observation_period.person_id'
* Missing field 'person.care_site_id'
* Missing field 'person.gender_concept_id'
* Missing field 'person.person_id'
* Missing field 'person.person_source_value'
* Missing field 'person.year_of_birth'
* Missing field 'procedure_occurrence.person_id'
* Missing field 'procedure_occurrence.procedure_concept_id'
* Missing field 'procedure_occurrence.procedure_date'
* Missing field 'procedure_occurrence.procedure_occurrence_id'
* Missing field 'procedure_occurrence.procedure_type_concept_id'
* Missing field 'provider.care_site_id'
* Missing field 'provider.provider_id'
* Missing field 'provider.provider_source_value'
* Missing field 'visit_occurrence.person_id'
* Missing field 'visit_occurrence.place_of_service_concept_id'
* Missing field 'visit_occurrence.visit_end_date'
* Missing field 'visit_occurrence.visit_occurrence_id'
* Missing field 'visit_occurrence.visit_start_date'
File steps.csv looks good!
File tools.csv looks good!
File sources.csv looks good!
File people.csv looks good!
Found 95 entities
Found 24 steps
Found 10 tools
Found 2 sources
Found 3 persons
```

**Comment by murphyke** [comment]
Byron, thanks.  I mis-documented Nemours, so its results are much better:

```
sh-3.2$ bin/pedsnet-etlprov -model i2b2_pedsnet -version 2.0.0 ~/Documents/pedsnet/Data_Coordinating_Center/provenance/data/nemours/
Validating against model 'i2b2_pedsnet/2.0.0'
A few problems have been detected with the entities.csv file:
* Missing field 'provider_dimension.provider_id'
* Missing field 'provider_dimension.provider_path'
File steps.csv looks good!
File tools.csv looks good!
File sources.csv looks good!
File people.csv looks good!
Found 95 entities
Found 24 steps
Found 10 tools
Found 2 sources
Found 3 persons
```

I have to look at those two fields in i2b2_pedsnet.

**Comment by gracecodes144** [comment]
This looks like it is already turning into a very useful tool. :+1:

**Comment by bruth** [comment]
@murphyke Updated report. It looks like someone copied someone.. (step 70)

**Nationwide**

```
Validating against model 'i2b2_pedsnet/2.0.0'
---
3 errors have been detected for '/Users/ruthb/Code/github.com/PEDSNet/Data_Coordinating_Center/provenance/data/nationwide/steps.csv'
* Error parsing previous step 'None'
* Entity 'provider_dimension' not defined
* Step 70 does not exist
---
2 entities are missing from the model
* provider_dimension.provider_id
* provider_dimension.provider_path
---
101 entities
24 steps
15 tools
2 sources
3 persons
```

**Nemours**

```
Validating against model 'i2b2_pedsnet/2.0.0'
---
8 errors have been detected for '/Users/ruthb/Code/github.com/PEDSNet/Data_Coordinating_Center/provenance/data/nemours/steps.csv'
* Error parsing previous step 'none'
* Entity 'patient_dimension.gestational_age_num' not defined
* Entity 'patient_dimension.gestational_age_num' not defined
* Entity 'visit_dimension.admit_src_destcd' not defined
* Entity 'patient_dimension.gestational_age_num' not defined
* Entity 'visit_dimension.admit_src_destcd' not defined
* Entity 'provider_dimension' not defined
* Step 70 does not exist
---
2 entities are missing from the model
* provider_dimension.provider_id
* provider_dimension.provider_path
---
95 entities
18 steps
10 tools
2 sources
3 persons
```

**Comment by bruth** [comment]
Edited the previous comment with the full error report.

**Comment by murphyke** [comment]
Thanks, @bruth

**Comment by bruth** [comment]
@murphyke Does the output satisfy this issue?

**Comment by murphyke** [comment]
Yes

**Comment by bruth** [comment]
Implemented in 9649bbb3c
